### PR TITLE
[PPS][NanoAOD] Generator-level proton information

### PR DESCRIPTION
**Pull request by forthommel**
#### PR description:

This PR introduces a new flat table for the storage of GEN-level proton properties at NanoAOD level. As described [in this cross-POG presentation](https://indico.cern.ch/event/1079178/contributions/4606748) the expected new collection size is at the order of 12-19 B/event, with a ~0.03 ms/event filling time.
The stored proton attributes are:
* its vertex longitudinal coordinate (to point at simulated vertex position)
* its transverse and longitudinal momentum (providing Mandelstam-t and the adimensional xi fractional momentum loss)
* a boolean flag stating if the proton arises from signal (e.g. central exclusive processes), or from pileup.

#### PR validation:

Code developed in 10_6_X, forward-ported to master branch, and tested in both environments. Compiles, produces the expected output ; matrix tests running.

#### if this PR is a backport please specify the original PR and why you need to backport that PR: N/A

Before submitting your pull requests, make sure you followed this checklist:
- [x] verify that the PR is really intended for the chosen branch
- [x] verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- [x] verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

cc: @michael-pitt, @fabferro

### CODE REVIEW

**Comment by mariadalfonso** [comment]
please test

**Comment by forthommel** [comment]
Failure in `1329.1` is a perfect example of an input file where the tag to the PU protons collection was attributed an extra `genPUProtons` postfix. We have a fix for these particular cases in https://github.com/cms-sw/cmssw/pull/36080/files#diff-7a4d1f8a67af14c00720d67b77419e6ae6d79974bb64881badf5ccf5da19bdfaR26, but unfortunately it seems to apply only for a certain class of samples (still did not find a common rule so far...)
Any idea on how to accommodate with this? Is there a particular (set of) era modifier(s) we can use?

**Comment by mariadalfonso** [comment]
> Failure in `1329.1` is a perfect example of an input file where the tag to the PU protons collection was attributed an extra `genPUProtons` postfix. We have a fix for these particular cases in https://github.com/cms-sw/cmssw/pull/36080/files#diff-7a4d1f8a67af14c00720d67b77419e6ae6d79974bb64881badf5ccf5da19bdfaR26, but unfortunately it seems to apply only for a certain class of samples (still did not find a common rule so far...)
> Any idea on how to accommodate with this? Is there a particular (set of) era modifier(s) we can use?

This postfix problem is isolated for the old miniAOD samples ? or can vary between samples of UL or Run3 ?

We already switched off the PPS for the old EOY nano, we can set of also these GenProtons

https://github.com/cms-sw/cmssw/pull/36080/files?authenticity_token=1VK9u43oz66LD3rvIX5ViQqnShItUvktU3i2kqwlbtWADeo7jtE4ZDoUqwgtZQ03OmFn8Lhxt%2F38RR5bbPjJMA%3D%3D&file-filters%5B%5D=.py#diff-7a4d1f8a67af14c00720d67b77419e6ae6d79974bb64881badf5ccf5da19bdfaR72-R73

**Comment by forthommel** [comment]
@mariadalfonso My bad, upon further investigation it seems that the input collection was not *at all* present in 8_0_X relvals. This postfix issue is therefore not involved in this failed matrix test (it seems it was only introduced in the 9_4_X cycle, see #20172).
Once I find the proper era modifier for these UL samples I will disable it in a later push. 
But to follow up on this collection name issue I am still tracking down the MC production campaign(s) that introduced the extra label.

**Comment by forthommel** [comment]
As previously discussed, a3c03d6 disables the `genProtonTablesTask` for `8_0_X` legacy samples.

File: PhysicsTools/NanoAOD/python/protons_cff.py [hunk 1]
```diff
@@ -19,6 +20,12 @@
 protonTable.tagRecoProtonsSingle = cms.InputTag("filteredProtons" if singleRPProtons else "ctppsProtons","singleRP")
 
 
+genProtonTable = _genproton.clone(
+    cut = cms.string('(pdgId == 2212) && (abs(pz) > 5200) && (abs(pz) < 6467.5)'), # xi in [0.015, 0.2]
+    # PU protons source could have an extra 'genPUProtons' collection name for some samples
+    # srcPUProtons = cms.InputTag('genPUProtons', 'genPUProtons'),
```
Comment (mariadalfonso) [inline PhysicsTools/NanoAOD/python/protons_cff.py:26]: are those comments line meaningful now ?
better to remove this commented code, if not needed

Comment (forthommel) [reply PhysicsTools/NanoAOD/python/protons_cff.py:26]: For the time being, as several privately-produced samples are relying on this patch we just left it for bookkeeping. If needed we can introduce a customisation command for `cmsDriver.py` instead?

**Comment by mariadalfonso** [comment]
please test

**Comment by forthommel** [comment]
The only 6 differences arise from the MessageLogger instance. This looks unspoiled...

**Comment by forthommel** [comment]
Hi @mariadalfonso, all,
Is there anything we can do to ease the review?

**Comment by forthommel** [comment]
Following the discussion at [yesterday's proton POG meeting](https://indico.cern.ch/event/1096937/contributions/4614858/) it was decided that both `px` and `py` were interesting to to be stored instead of `pt` only. Furthermore, I took the opportunity to rebase from `pre1` to `pre2`.

**Comment by gouskos** [comment]
please test

**Comment by forthommel** [comment]
Here are the average `GenProton` table sizes (5-6k events each).

* inclusive sample [1] (PU protons only, no signal outgoing protons):
  Screenshot from 2021-11-17 14-58-48
* exclusive sample [2] (signal + PU protons):
  Screenshot from 2021-11-17 14-59-27

[1] `/store/mc/RunIISummer20UL16MiniAODAPVv2/DYJetsToLL_M-50_TuneCP5_13TeV-madgraphMLM-pythia8/MINIAODSIM/106X_mcRun2_asymptotic_preVFP_v11-v1/280000/A163995F-0B43-6749-A2B5-3256F7CC714D.root`
[2] `/store/mc/RunIISummer20UL16MiniAODAPVv2/GGToMuMu_Pt-25_Elastic_13TeV-lpair/MINIAODSIM/106X_mcRun2_asymptotic_preVFP_v11-v1/280000/3870E880-4A47-7440-B122-C76062D2290F.root`

**Comment by forthommel** [comment]
@mariadalfonso, @gouskos, all,
Is there anything we can provide to make your review easier?

**Comment by gouskos** [comment]
Thanks @forthommel !
There is an error [i.e., https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/17773169] when running over the 94X miniAOD samples. Could you please fix it in the same manner as you did it for 80X in protons_cff.py

**Comment by forthommel** [comment]
Thanks to you, @gouskos!
Indeed, I did not spot it with my recipe. For my personal information, could you please point me at the recipe for the production of this `TTToSemiLeptonic_TuneCP5_PSweights_13TeV-powheg-pythia8-100ev.root` input file to reproduce the environment?

**Comment by forthommel** [comment]
Nevermind, I found `/store/cmst3/group/nanoAOD/pre-94XMiniAODv2/TTToSemiLeptonic_TuneCP5_PSweights_13TeV-powheg-pythia8-100ev.root` in [the CI definitions](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/blob/master/scripts/run_test_simple.sh#L29). It appears the `genPUProtons` collection (with the extra `genPUProtons` label) was indeed present for this 94X production. I will continue my tracking down task to find out when this extra label was removed and propose era modifier(s) accordingly).

**Comment by gouskos** [comment]
Here it is the cmsDriver command:
https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/blob/master/scripts/run_test_simple.sh#L30
I do not think you can access the MiniAOD inputs file. If you do not have any 94X handy, please let me know and I will copy it

**Comment by forthommel** [comment]
Thanks @gouskos. Yes actually, I have read permission to the T3, so I should be able to reproduce the environment.

**Comment by forthommel** [comment]
Interesting! On my test bench everything runs smoothly with a combination of this input file and
`cmsDriver.py test_mc_94Xv2 -s NANO --mc --eventcontent NANOAODSIM --datatier NANOAODSIM --filein file:TTToSemiLeptonic_TuneCP5_PSweights_13TeV-powheg-pythia8-100ev.root --conditions auto:phase1_2017_realistic -n 100 --era Run2_2017,run2_nanoAOD_94XMiniAODv2 | tee $OUTPUTS/simple_test_mc_94Xv2.txt && cp -v test_mc_94Xv2_NANO.root $OUTPUTS/ && ${sanity_check_py_ind} -r 1 -e 100 $OUTPUTS/test_mc_94Xv2_NANO.root && echo '- Test mc_94Xv2: passed' >>  $OUTPUTS/tests-simple.txt` on my 12_2_0_pre2.

**Comment by forthommel** [comment]
My bad, I already introduced the modification elsewhere in my testing environment. 
810133d modifies the input label for 94X-produced (legacy run 2) MiniAOD samples.

**Comment by gouskos** [comment]
Thanks @forthommel !
I already launched the dedicated nanoAOD tests - I ll keep you posted

**Comment by gouskos** [comment]
please test

**Comment by forthommel** [comment]
[Long tests](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/17784786) show that there is also a correction required for nanoAOD DQM part. Any hint and/or guidance?

**Comment by forthommel** [comment]
@gouskos 170fa71 fixes both @cmsbuild tests and the cms-nanoAOD one.

**Comment by gouskos** [comment]
please test

**Comment by forthommel** [comment]
Local unit test raises a FileOpenError for runtestPhysicsToolsNanoAOD. 
`/store/relval/CMSSW_9_4_0_pre3/RelValTTbar_13/MINIAODSIM/PU25ns_94X_mc2017_realistic_v4-v1/10000/52B94CC0-6FBB-E711-B577-0CC47A7C35F8.root` cannot be opened for [94X v1 test](https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/test/runtests.sh#L9). Not clear if this is related to this unexpected last minute failure.

**Comment by gouskos** [comment]
> Local unit test raises a FileOpenError for runtestPhysicsToolsNanoAOD. `/store/relval/CMSSW_9_4_0_pre3/RelValTTbar_13/MINIAODSIM/PU25ns_94X_mc2017_realistic_v4-v1/10000/52B94CC0-6FBB-E711-B577-0CC47A7C35F8.root` cannot be opened for [94X v1 test](https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/test/runtests.sh#L9). Not clear if this is related to this unexpected last minute failure.

Just saw it @forthommel - Will check this.

**Comment by forthommel** [comment]
Furthermore, my local test did not pick up the issue in `1325.7`. It seems that the extra `genPUProtons` was really scattered all over 94X eras. Two possible fixes would be
1) disable the module for all pre-10XY eras (hence leaving the user to tweak its configuration for a re-processing of legacy centrally-produced 94X miniAOD samples to add this extra table), or
2) introduce an automated fallback retrieval of a `edm::InputTag("genPUProtons", "genPUProtons")`-labeled collection if the `edm::InputTag("genPUProtons", "")` one was not found in event content.
Which of these proposals seem more reasonable?

(adding @antoniovilela to the loop as author of https://github.com/cms-sw/cmssw/pull/20172 and who might be able to comment on these various InputTag namings...)

**Comment by antoniovilela** [comment]
> Furthermore, my local test did not pick up the issue in `1325.7`. It seems that the extra `genPUProtons` was really scattered all over 94X eras. Two possible fixes would be
> 
> 1. disable the module for all pre-10XY eras (hence leaving the user to tweak its configuration for a re-processing of legacy centrally-produced 94X miniAOD samples to add this extra table), or
> 2. introduce an automated fallback retrieval of a `edm::InputTag("genPUProtons", "genPUProtons")`-labeled collection if the `edm::InputTag("genPUProtons", "")` one was not found in event content.
>    Which of these proposals seem more reasonable?
> 
> (adding @antoniovilela to the loop as author of #20172 and who might be able to comment on these various InputTag namings...)

I need to go back and check (I cannot do it today), but there was a difference whether it was a standard mixing or pre-mixing workflow.

**Comment by mariadalfonso** [comment]
please test

**Comment by forthommel** [comment]
Good! Are we ready for a review, leaving @antoniovilela's possible comments/suggestions for this mix/premix tags namings to a potential follow-up PR?

**Comment by mariadalfonso** [comment]
> Good! Are we ready for a review, leaving @antoniovilela's possible comments/suggestions for this mix/premix tags namings to a potential follow-up PR?

still some failure: you should disable also the 94X2016,102X.
In this way will remain enabled only for the  UL

**Comment by forthommel** [comment]
@mariadalfonso @gouskos
Friendly reminder of the existence of this PR :)

**Comment by mariadalfonso** [comment]
@forthommel

there are still failures, see here
https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/18098699

**Comment by forthommel** [comment]
Thanks, @mariadalfonso ! This was indeed unnoticed. I believe we are going to test disabling this module over every single era until we properly understand the naming of this collection...
@antoniovilela any news on your investigation?

**Comment by antoniovilela** [comment]
> Thanks, @mariadalfonso ! This was indeed unnoticed. I believe we are going to test disabling this module over every single era until we properly understand the naming of this collection... @antoniovilela any news on your investigation?

It seems indeed that with the current behavior you still need to define the input collection as "cms.InputTag("genPUProtons","genPUProtons")" when using a sample that was simulated with pre-mixing. This appears in the "mixData" module which makes a copy of the original collection.
One could presumably make the collection naming uniform, either removing the instance label in the pre-mixing workflow, or adding it in the original producer which runs with standard mixing and when producing the premix dataset, but this is the current behavior.

**Comment by forthommel** [comment]
As this PR was one of the "overlapping ones", happy new year to y'all! Hope these wishes (and a few commits) will allow us to converge on this topic!

@antoniovilela Many thanks for your check! True, having two collections labels for the same info seems (at least to me) a bit untrivial to manage. Do you have a patch proposal in mind or shall I investigate myself?

@mariadalfonso, @gouskos, be038f3 was (force-)pushed after its rebase to `10_3_X` to accommodate two possible input collection labels. Possibly, other implementations of such a "fallback collection retrieval" can be found elsewhere in production code, so maybe you are aware of anything more efficient..?

**Comment by forthommel** [comment]
Thanks for triggering the tests, @gouskos ! Along with @mariadalfonso, were the nanoAOD "private" tests already triggered?

**Comment by gouskos** [comment]
Thanks @forthommel The "private" tests are done. Checking the outputs and we will come back to you very soon

**Comment by forthommel** [comment]
@gouskos Could you please let us know what is the status of your checks?

**Comment by gouskos** [comment]
Hi @forthommel:
- all of our "private" tests [https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/issues/135#note_5156614]
- We see ~3% increase in the size for UL16 and UL18, which we should try to reduce. Can you please reduce the precision of the stored variables to match the one used for storing the genParticles [e.g.,: https://github.com/cms-sw/cmssw/blob/f904c8155ac865af2a57381cfda7ccfa9ea4337c/PhysicsTools/NanoAOD/python/genparticles_cff.py#L41]

**Comment by forthommel** [comment]
Hi @gouskos,
In 1f74f4b I implemented a mantissa reduction (precision=8) for all single floating point variables. With respect to the &gamma;&gamma; &rarr; &mu;<sup>+</sup>&mu;<sup>-</sup> signal used in [the former benchmarking](https://github.com/cms-sw/cmssw/pull/36080#issuecomment-971612321) it results in roughly 50% reduction (44 → 21 bytes/event) in output size for the new collection.
image

**Comment by gouskos** [comment]
please test

**Comment by gouskos** [comment]
Thanks @forthommel !
Exactly, ~50% is what I would expect.

> Hi @gouskos,
> In [1f74f4b](https://github.com/cms-sw/cmssw/commit/1f74f4b2bf423897b4b647b35c5eaff4702325e9) I implemented a mantissa reduction (precision=8) for all single floating point variables. With respect to the γγ → μ+μ- signal used in [the former benchmarking](https://github.com/cms-sw/cmssw/pull/36080#issuecomment-971612321) it results in roughly 50% reduction (44 → 21 bytes/event) in output size for the new collection.
> image

**Comment by forthommel** [comment]
No obvious reason why this latter patch would spontaneously fail on a 2012 ZMu skim... Currently running it locally.

**Comment by forthommel** [comment]
Indeed, a good old ```XrdCl::File::Open(name='root://cms-xrd-global.cern.ch//store/data/Run2012D/SingleMu/RAW-RECO/ZMu-15Apr2014-v1/00000/0C402854-EECC-E311-AE34-E0CB4E553656.root', flags=0x10, permissions=0660) => error '[ERROR] Server responded with an error: [3011] No servers are available to read the file.``` input file error...

**Comment by grzanka** [comment]
Is this PR still waiting for this https://github.com/cms-sw/cmssw/issues/36771 to be resolved ?

**Comment by perrotta** [comment]
> Is this PR still waiting for this #36771 to be resolved ?

No, just forget about that workflow for the moment

**Comment by grzanka** [comment]
@perrotta it seems https://github.com/cms-sw/cmssw/issues/36771 is fixed now

**Comment by perrotta** [comment]
please test
(just to refresh them)

**Comment by mariadalfonso** [comment]
@forthommel
do you plan to update to further reduce the size ?

**Comment by mariadalfonso** [comment]
@forthommel ping

**Comment by forthommel** [comment]
Hello @mariadalfonso, @gouskos, and thanks for your patience!
Together with @michael-pitt we studied the possible reduction in the overall footprint of this new collection. In particular, we left the mantissa bits multiplicity as a free parameter, and observed the effect of the precision drop as translated into variables of interest for PPS/forward physics studies.
These variables are namely, the proton usual p<sub>T</sub>, the longitudinal momentum loss &xi; = &Delta;p<sub>z</sub>/p<sub>z</sub>, and the longitudinal coordinate of the scattered proton vertex.
From the plot below, we observe that a 5-bit mantissa would be enough to achieve a MeV-scale precision in the proton p<sub>T</sub>. For the most significant variable &xi; we would leave it unchanged to a 8-bit mantissa to achieve a sub-percent absolute precision, and a precision-5 or 6 would work for the proton vertex longitudinal coordinate.
image
In addition, here is a numbered summary of the added variable size (in byte/event) as a function of this mantissa dynamics:
| Variable | n=4 | n=5 | n=6 | n=8 | uncons. |
|------------|-------|-------|-------|-------|------------|
| `GenProton_px` | 4.8 | 5.2 | 5.4 | 6.1 | 11.6 |
| `GenProton_py` | 4.8 | 5.2 | 5.5 | 6.0 | 11.6 |
| `GenProton_pz` | 2.5 | 2.9 | 3.3 | 4.0 | 9.9 |
| `GenProton_vz` | 4.6 | 5.0 | 5.2 | 5.8 | 10.7 |

Hence, given the figures derived above, we would go from a ~21.9 to 19.4-19.6 b/event for the floating-point variables only (in addition to the fixed 2.1 b/event for the indexing integer/`GenProton_isPU` boolean).
@michael-pitt can possibly comment on the effect of further constraining the proton selection cut.

**Comment by gouskos** [comment]
Hi @forthommel and @michael-pitt thank you very much for the detailed and very useful studies!
I think we can stick with the current precision [=8]
Let me rerun the NanoAOD checks ones more and then we merge

> Hello @mariadalfonso, @gouskos, and thanks for your patience! Together with @michael-pitt we studied the possible reduction in the overall footprint of this new collection. In particular, we left the mantissa bits multiplicity as a free parameter, and observed the effect of the precision drop as translated into variables of interest for PPS/forward physics studies. These variables are namely, the proton usual pT, the longitudinal momentum loss ξ = Δpz/pz, and the longitudinal coordinate of the scattered proton vertex. From the plot below, we observe that a 5-bit mantissa would be enough to achieve a MeV-scale precision in the proton pT. For the most significant variable ξ we would leave it unchanged to a 8-bit mantissa to achieve a sub-percent absolute precision, and a precision-5 or 6 would work for the proton vertex longitudinal coordinate. image In addition, here is a numbered summary of the added variable size (in byte/event) as a function of this mantissa dynamics:
> 
> Variable	n=4	n=5	n=6	n=8	uncons.
> `GenProton_px`	4.8	5.2	5.4	6.1	11.6
> `GenProton_py`	4.8	5.2	5.5	6.0	11.6
> `GenProton_pz`	2.5	2.9	3.3	4.0	9.9
> `GenProton_vz`	4.6	5.0	5.2	5.8	10.7
> Hence, given the figures derived above, we would go from a ~21.9 to 19.4-19.6 b/event for the floating-point variables only (in addition to the fixed 2.1 b/event for the indexing integer/`GenProton_isPU` boolean). @michael-pitt can possibly comment on the effect of further constraining the proton selection cut.

**Comment by forthommel** [comment]
[It looks like](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/issues/135#note_5273080) the NanoAOD checks yielded a comparable O(20 b/ev) increase in 10_6_X MC workflows.

**Comment by forthommel** [comment]
Many thanks, @perrotta !

**Pull request merged**